### DISCUSSION TITLE
How to let Knowledge base works ignore upper/lower case

### DISCUSSION THREAD

**Discussion by aoleiReiz** [explosion/spaCy#7801] [Help: Best practices]
currently I have kb which all upper cased, it Entity linker failed to get kb_id when do prediction on lowered text. could you please help how to solve it

**Comment by svlandeg** [comment]
Hi! 

From spaCy v3.0 onwards, it should (hopefully) be relatively straightforward to tune the candidate generation to your needs. The candidate generation step selects all relevant IDs from the Knowledge base, given a certain textual mention. More specifically, it needs to be a function that takes a `KnowledgeBase` and a `Span`, and outputs a list of `Candidate` instances:

The default candidate generation method is `spacy.CandidateGenerator.v1` and you are correct that it works case-sensitive:
```
@registry.misc("spacy.CandidateGenerator.v1")
def create_candidates():
    return get_candidates

def get_candidates(kb, span):
    return kb.get_alias_candidates(span.text)
```
You should be able to overwrite this behaviour by defining your own candidate generator, e.g. something like this:
```
@registry.misc("custom_upper_candidates")
def create_candidates():
    return get_custom_candidates

def get_custom_candidates(kb, span):
    cased_results = kb.get_alias_candidates(span.text)
    if len(cased_results) > 0:
        return cased_results
    return kb.get_alias_candidates(span.text.upper())
```
And then you'd define this candidate generator in your config file:
```
[components.entity_linker]
factory = "entity_linker"
entity_vector_length = 64
get_candidates = {"@misc":"custom_upper_candidates"}
incl_context = true
incl_prior = true
labels_discard = []
```
Let us know if you run into issues with that!

**Reply by amirj** [reply]
Thanks @svlandeg 

Assume that we're creating a more complex candidate generator function; we need to return a list of candidates as follow:

```python
def get_custom_candidates(kb, span):
    candidate = Candidate(kb, entity_hash, entity_freq, entity_vector, alias_hash, prior_prob)
    return [candidate]
```

In order to create a [Candidate](https://spacy.io/api/kb#candidate) object we need `entity_hash, entity_freq, entity_vector, alias_hash, prior_prob`. To the best of my knowledge, some of these items (`entity_vector , prior_prob`) can be retrieved from kb, but it's not clear to me how to assign `entity_hash, entity_freq, alias_hash`?

**Reply by svlandeg** [reply]
These are closely tied to the other methods in the [`KnowledgeBase`](https://spacy.io/api/kb) as implemented internally in spaCy, so you can find the definitions in the documentation:

`entity_freq` and `entity_vector` are the two arguments you'd pass when calling [`add_entity`](https://spacy.io/api/kb#add_entity). `entity_freq` is currently not used by the implementation of the `entity_linker`.

`prior_prob` is the probability you'd pass on for a given entity-alias combination when calling [`add_alias`](https://spacy.io/api/kb#add_alias).

`entity_hash` and `alias_hash` are the hashes of the entity ID and the mention string in the vocab respectively, defined by:
```
# when adding the strings for the first time
entity_hash = self.vocab.strings.add(entity)
alias_hash = self.vocab.strings.add(alias)
```
or
```
# assuming the strings are already in the vocab
entity_hash = self.vocab.strings[entity]
alias_hash = self.vocab.strings[alias]
```

**Reply by amirj** [reply]
Thank @svlandeg 

I'm using the following functions:

```python
@registry.misc("custom_candidate_generator")
def create_candidates():
    return get_custom_candidates


def get_custom_candidates(kb, span):
    candidates = []
    for ent_id, score in candidate_generator_func():
        candidate = Candidate(
            kb=kb,
            entity_hash=kb.vocab.strings[ent_id],
            entity_freq=111,
            entity_vector=kb.get_vector(ent_id),
            alias_hash=kb.vocab.strings[span.text],
            prior_prob=score,
        )
        candidates.append(candidate)
    return candidates

```

Add entity linker pipe and initialize for training:

```python
config = {
    "labels_discard": [],
    "n_sents": 0,
    "incl_prior": True,
    "incl_context": True,
    "model": DEFAULT_NEL_MODEL,
    "entity_vector_length": 200,
    "get_candidates": {"@misc": "custom_candidate_generator"},
}
el = self.nlp.add_pipe("entity_linker", config=config, last=True)
el.initialize(get_examples=self.get_examples, kb_loader=load_kb(loc_kb)) #loc_kb is the Path of KB on disk
```

EntityLinker is trained for one epoch, after that when I'm going to evaluate the model on a dev set, I'm getting the following error:

```
***/python3.8/site-packages/spacy/pipeline/entity_linker.py:362: RuntimeWarning: invalid value encountered in true_divide
  sims = xp.dot(
Traceback (most recent call last):
  File "sample_entity_linker/models/spacy_entity_linker.py", line 292, in <module>
    spacy_entity_linker.train(epoches=10)
  File "sample_entity_linker/models/spacy_entity_linker.py", line 205, in train
    score = self.evaluate(dataset=self.dev_set)
  File "sample_entity_linker/models/spacy_entity_linker.py", line 220, in evaluate
    score = self.nlp.evaluate(dataset)
  File "/home/aj896572/.conda/envs/sample-entity-linker/lib/python3.8/site-packages/spacy/language.py", line 1355, in evaluate
    for doc, eg in zip(
  File "/home/aj896572/.conda/envs/sample-entity-linker/lib/python3.8/site-packages/spacy/util.py", line 1484, in _pipe
    yield from proc.pipe(docs, **kwargs)
  File "spacy/pipeline/trainable_pipe.pyx", line 79, in pipe
  File "/home/aj896572/.conda/envs/sample-entity-linker/lib/python3.8/site-packages/spacy/util.py", line 1503, in raise_error
    raise e
  File "spacy/pipeline/trainable_pipe.pyx", line 75, in spacy.pipeline.trainable_pipe.TrainablePipe.pipe
  File "/home/aj896572/.conda/envs/sample-entity-linker/lib/python3.8/site-packages/spacy/pipeline/entity_linker.py", line 373, in predict
    final_kb_ids.append(best_candidate.entity_)
  File "spacy/kb.pyx", line 45, in spacy.kb.Candidate.entity_.__get__
  File "spacy/strings.pyx", line 132, in spacy.strings.StringStore.__getitem__
KeyError: "[E018] Can't retrieve string for hash '17224167776089022530'. This usually refers to an issue with the `Vocab` or `StringStore`."
```

I'm not sure to completely understand the relationship between EntityLinker and KnowledgeBase, i.e. when adding a new entity_linker pipe, it's pairing to which kb?! given the fact that we call "initialize" function only for training purpose --I see a `set_kb` function in the documentation but I'm not using it in my code!

**Reply by svlandeg** [reply]
This might be partially addressed by https://github.com/explosion/spaCy/discussions/8080, but I wanted to come back to this question:
> when adding a new entity_linker pipe, it's pairing to which kb?! 

Upon construction, the `entity_linker` [gets](https://github.com/explosion/spaCy/blob/d2bdaa78231b62ae06afbd8ad92a22e0cd9cf1dd/spacy/pipeline/entity_linker.py#L147) an empty `KnowledgeBase`. You need to override this with either `set_kb` or `initialize`.
